src/main/java/org/folio/rest/impl/InstanceStorageAPI.java

```diff
@@ -118,29 +123,56 @@ public void postInstanceStorageInstances(
             }
           }
 
-          postgresClient.save(INSTANCE_TABLE, entity.getId(), entity,
-            reply -> {
-              try {
-                if(reply.succeeded()) {
-                  asyncResultHandler.handle(
-                    io.vertx.core.Future.succeededFuture(
-                      PostInstanceStorageInstancesResponse
-                        .respond201WithApplicationJson(entity,PostInstanceStorageInstancesResponse.headersFor201().withLocation(reply.result()))));
-                }
-                else {
+          final Future<String> hridFuture;
+          if (isBlank(entity.getHrid())) {
+            final HridManager hridManager = new HridManager(vertxContext, postgresClient);
+            hridFuture = hridManager.getNextInstanceHrid();
+          } else {
+            hridFuture = Promise.succeededPromise(entity.getHrid()).future();
+          }
+
+          hridFuture.map(hrid -> {
+            entity.setHrid(hrid);
+            postgresClient.save(INSTANCE_TABLE, entity.getId(), entity,
+              reply -> {
+                try {
+                  if(reply.succeeded()) {
+                    asyncResultHandler.handle(
+                      io.vertx.core.Future.succeededFuture(
+                        PostInstanceStorageInstancesResponse
+                          .respond201WithApplicationJson(entity,
+                              PostInstanceStorageInstancesResponse.headersFor201().withLocation(reply.result()))));
+                  }
+                  else {
+                    if (PgExceptionUtil.isUniqueViolation(reply.cause())) {
+                      asyncResultHandler.handle(
+                          io.vertx.core.Future.succeededFuture(
+                            PostInstanceStorageInstancesResponse
+                              .respond400WithTextPlain(PgExceptionUtil.badRequestMessage(reply.cause()))));
```

Comment:
Agreed. The problem here is that the `instance-storage` API does not use the validation RAML trait and therefore is not supposed to return 422 (no `respond422XXX` method in the interface). Of course, we could create our own response, but again, it would be unexpected. Since this is a major, at least in my mind, change to the interface, I was reluctant to make this change. I think we need a separate story (which we do already have, in fact, since I created it about 2 years ago 😈) to add/support validation to all mod-inventory-storage APIs (some other APIs have validation enabled but do not have the Validate annotation on the impl methods in order for validation to work).